[Profiler x RPC] Enable RPC Server Global Profiler

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38847 [Profiler x RPC] Enable RPC Server Global Profiler**

See motivation and design in https://github.com/pytorch/pytorch/issues/38845.

Close https://github.com/pytorch/pytorch/issues/38845.

Changes,

- Add pre-request and post-response hooks to RPC "request_callback_impl.cpp". For a thread that executes RPC handler, check if the server-side global profiling is on. If it's on, enable profiling on this thread and after response, merge the thread-local profiling result into the global profiling state.
- Add context-style Python API to parse the profiling Events into ranges represented by FunctionEvent.
- Add data-structures to work as global profiling state that support nesting and container for consolidating results from multiple threads.

Test,

- Add a test that uses nested profiling range and inspect the profiling events.

Differential Revision: [D5665992](https://our.internmc.facebook.com/intern/diff/D5665992/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5665992/)!

xush6528 committed:
[Profiler x RPC] Enable RPC Server Global Profiler

See plan in https://github.com/pytorch/pytorch/issues/38845

Differential Revision: [D5665992](https://our.internmc.facebook.com/intern/diff/D5665992/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5665992/)!

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -976,6 +981,35 @@ def test_profiler_with_script_remote_rpc(self):
+
+        rpc.rpc_sync(dst_worker_name, enable_profiler)
+        rpc.rpc_sync(dst_worker_name, torch.add, (x, y))
+        events = rpc.rpc_sync(dst_worker_name, rpc.disable_server_process_global_profiler)
```
What happens in the case where multiple workers send RPCs to the server at the same time? Are we guaranteed that only profiling results returned to a worker will be due to work they initiated on the server? Will this invoke the overhead of profiling other workers' requests even if they have not specifically requested for that?

xush6528 replied:
@rohan-varma 
No matter which client worker or whether the client enables profiling or not, all threads that handle requests do profiling.
That's why I am calling it server global profiling.

This overhead is ok for our use case. When we enable this on an embedding server, all the RPCs the server do is embedding lookup forward and backward passes. If there is more types of RPCs involved like signaling/metric reporting, we can optimize those later.

xush6528 committed:
Update on "[Profiler x RPC] Enable RPC Server Global Profiler"


See plan in https://github.com/pytorch/pytorch/issues/38845

Differential Revision: [D5665992](https://our.internmc.facebook.com/intern/diff/D5665992/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5665992/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[Profiler x RPC] Enable RPC Server Global Profiler"


See plan in https://github.com/pytorch/pytorch/issues/38845

Differential Revision: [D5665992](https://our.internmc.facebook.com/intern/diff/D5665992/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5665992/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[Profiler x RPC] Enable RPC Server Global Profiler"


See plan in https://github.com/pytorch/pytorch/issues/38845

Differential Revision: [D5665992](https://our.internmc.facebook.com/intern/diff/D5665992/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5665992/)!

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -578,6 +579,27 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processMessage(
+          if (isServerProcessGlobalProfilerEnabled) {
+            // Initialize thread-local profiler state from process-global
+            // profiler state.
+            serverProcessGlobalProfilerStatePtr =
```
This is in the RRef ready callback, which can run on a different thread. Is this intentional?

xush6528 replied:
Yes, intentional. It doesn't matter.

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -578,6 +579,27 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processMessage(
+          // of 10us. If server global profiler is enabled, we futher pay the
+          // cost of thread local profiler state initialization .
+          // int64_t startTime = ::torch::autograd::profiler::getTime();
+          bool isServerProcessGlobalProfilerEnabled =
```
This means profiler will never care about the time spent on waiting for RRef confirmations?

xush6528 replied:
Profiler only take into account Tensor "function"s and user specified `RecordFunction` scope. It's not aware of RPC time by default.

mrshenli commented on torch/csrc/distributed/rpc/profiler/server_process_global_profiler.cpp:
```diff
@@ -0,0 +1,115 @@
+std::mutex
+#else
+std::shared_timed_mutex
+#endif
```
I wonder if we should use `#if` in conjunction with `using` to define a mutex type, so that we don't have to do this four times.

xush6528 replied:
Good idea. Updating.

mrshenli commented on torch/csrc/distributed/rpc/profiler/server_process_global_profiler.cpp:
```diff
@@ -0,0 +1,115 @@
+#endif
+        wlock(currentProcessGlobalProfilerStateStackEntryMutex);
+
+    auto previousStateStackEntryPtr =
+        currentProcessGlobalProfilerStateStackEntryPtr;
+    currentProcessGlobalProfilerStateStackEntryPtr =
+        std::make_shared<ProcessGlobalProfilerStateStackEntry>();
+    currentProcessGlobalProfilerStateStackEntryPtr
+        ->previousProcessGlobalProfilerStateStackEntryPtr_ =
+        previousStateStackEntryPtr;
+    currentProcessGlobalProfilerStateStackEntryPtr
+        ->profilerProcessGlobalStatePtr_ =
```
Is there any standard way to make good acronyms for variable names? If ProcessGlobalProfiler is widely used in our code base, shall we just call it `Pgp` for abbreviation? :)

xush6528 replied:
Let me put them into namespace to remove common prefix.

mrshenli commented on torch/csrc/distributed/rpc/profiler/server_process_global_profiler.h:
```diff
@@ -0,0 +1,66 @@
+
+using namespace torch::autograd::profiler;
+
+// Process global profiler state.
```
Shall we expand the comment to explain why we need this class, and where it is used?

xush6528 replied:
Adding.

mrshenli commented on torch/distributed/rpc/server_process_global_profiiler.py:
```diff
@@ -0,0 +1,34 @@
+    _enable_server_process_global_profiler(config)
+
+
+def disable_server_process_global_profiler():
```
are these user-facing APIs? If so, let's add docs?

mrshenli commented on torch/distributed/rpc/server_process_global_profiiler.py:
```diff
@@ -0,0 +1,34 @@
+        ]
+    )
+
+    return torch.autograd.profiler.EventList(
```
Is it intentional to make `disable_server_process_global_profiler` the only way to retrieve the event list? What if I'd like to retrieve the list but don't yet want to disable it?

xush6528 replied:
Yes, it should be the only way. It's the counterpart of autograd profiler, disableProfiler, API. I don't think there is other use cases.

xush6528 committed:
Update on "[Request for Comments] [Profiler x RPC] Enable RPC Server Global Profiler"


See plan in https://github.com/pytorch/pytorch/issues/38845

Differential Revision: [D5665992](https://our.internmc.facebook.com/intern/diff/D5665992/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5665992/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[Profiler x RPC] Enable RPC Server Global Profiler"


See plan in https://github.com/pytorch/pytorch/issues/38845

Differential Revision: [D5665992](https://our.internmc.facebook.com/intern/diff/D5665992/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5665992/)!

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/profiler/server_process_global_profiler.cpp:
```diff
@@ -0,0 +1,96 @@
+#include <shared_mutex>
+
+#include <torch/csrc/distributed/rpc/profiler/server_process_global_profiler.h>
```
Looks like other cpp files add its own header as the first include?

xush6528 replied:
Moving.

mrshenli commented on torch/csrc/distributed/rpc/profiler/server_process_global_profiler.cpp:
```diff
@@ -0,0 +1,96 @@
+using wLockType = std::unique_lock<std::mutex>;
+#else
+using mutexType = std::shared_timed_mutex;
+using rLockType = std::shared_lock<std::shared_timed_mutex>;
+using wLockType = std::unique_lock<std::shared_timed_mutex>;
```
macOS uses two unique_locks but other platforms use a shared_lock + a unique_lock. Is this intentional. If yes, why?

mrshenli replied:
If these are read-write locks, the read lock should be shared_lock as well for mac?

xush6528 replied:
The macOS versiono PyTorch CI uses is rather old.
`shared_lock` is not avaialble.

It shows this error
```
// Compiler error: 'shared_lock' is unavailable: introduced in
// macOS 10.12
```
We can remove this macro branching only after we abandon old macOS support.

mrshenli replied:
Let's keep this as-is. IIUC, there is no correctness problem here, and just a little slower if not using shared_lock for rlock, right?

xush6528 replied:
It's a bit slower for macOS. But if a user is using macOS for distributed training, I guess it's not a real use case in my understanding.

mrshenli commented on torch/csrc/distributed/rpc/profiler/server_process_global_profiler.cpp:
```diff
@@ -0,0 +1,96 @@
+
+void enableServer(const ProfilerConfig& new_config) {
+  TORCH_CHECK(
+      new_config.state == ProfilerState::CPU, "RPC module only support CPU");
```
CPU -> profiling CPU state?

As we will be adding CUDA support through TensorPipe in v1.6

xush6528 replied:
Let me remove this assertion. CPU+GPU training is supported with TensorPipe.

mrshenli commented on torch/csrc/distributed/rpc/profiler/server_process_global_profiler.cpp:
```diff
@@ -0,0 +1,96 @@
+  TORCH_CHECK(
+      new_config.state == ProfilerState::CPU, "RPC module only support CPU");
+  auto new_state = std::make_shared<State>(new_config);
+  StateStackEntry::push(new_state);
```
std::move?

xush6528 replied:
Adding.

mrshenli commented on torch/csrc/distributed/rpc/profiler/server_process_global_profiler.h:
```diff
@@ -0,0 +1,91 @@
+//
+// This class holds information about a profiling range, from "enable" to
+// "disable".
+// A instance of this ``State`` will be
```
A -> An

mrshenli commented on torch/csrc/distributed/rpc/profiler/server_process_global_profiler.h:
```diff
@@ -0,0 +1,91 @@
+    std::unique_lock<std::mutex> lock(profileRangesMutex_);
+
+    std::vector<thread_event_lists> results;
+    results.insert(
+        results.begin(),
+        std::make_move_iterator(profileRanges_.begin()),
+        std::make_move_iterator(profileRanges_.end()));
+    profileRanges_.erase(profileRanges_.begin(), profileRanges_.end());
```
can we simply swap these two vectors?

xush6528 replied:
Yes, that's better.

mrshenli commented on torch/csrc/distributed/rpc/profiler/server_process_global_profiler.h:
```diff
@@ -0,0 +1,91 @@
+  }
+
+ private:
+  // Name it profileRanges_ to emphesize on the fact that each element is the
```
emphasize

xush6528 replied:
Updating

mrshenli commented on torch/csrc/distributed/rpc/profiler/server_process_global_profiler.h:
```diff
@@ -0,0 +1,91 @@
+ private:
+  // Name it profileRanges_ to emphesize on the fact that each element is the
+  // results of a profile range. In each profile range, there is a
+  // "__profiler_start" mark event that all following events calculate time
```
mark -> marker?

xush6528 replied:
Yes.

mrshenli commented on torch/csrc/distributed/rpc/profiler/server_process_global_profiler.h:
```diff
@@ -0,0 +1,91 @@
+  // parse_cpu_trace(profileRange) for every profile range.
+  std::mutex profileRangesMutex_;
+  std::vector<thread_event_lists> profileRanges_;
+  ProfilerConfig config_ =
```
The ctor also sets `config_`. Is it necessary to provide default value for it? Looks like it will always be overwritten?

If we let ctor do the only init, this can be a `const`?

xush6528 replied:
It works. Updating.

mrshenli commented on torch/csrc/distributed/rpc/profiler/server_process_global_profiler.h:
```diff
@@ -0,0 +1,91 @@
+// on. This enables all RPC threads running server-side request callbacks.
+TORCH_API void enableServer(const ProfilerConfig& new_config);
+//
+// Turn on the state that indicates the server-side process-global profiling is
```
on -> off
enables -> disables?

xush6528 replied:
Updating.

mrshenli commented on torch/csrc/distributed/rpc/profiler/server_process_global_profiler.h:
```diff
@@ -0,0 +1,91 @@
+//
+// Turn on the state that indicates the server-side process-global profiling is
+// on. This enables all RPC threads running server-side request callbacks.
+TORCH_API std::vector<thread_event_lists> disableServer();
```
Actually, if the comment above true? Looks like `disableServer` only pops one from the stack, and as long as there are remaining elements in the stack, is it still considered enabled?

xush6528 replied:
Rephrasing it to express the nesting behavior.

mrshenli commented on torch/distributed/rpc/server_process_global_profiler.py:
```diff
@@ -0,0 +1,57 @@
+    _enable_server_process_global_profiler(config)
+
+
+def disable_server_process_global_profiler():
```
IIUC, the current version could return more events than those that are issued by the current worker? And the this worker needs to somehow filter those events to get relevant ones? If so, shall we make these two APIs private for now (starts with "_" and also need to address name conflict)? Because it doesn't seem very user-friendly. We might need more helpers/examples/tutorials for these APIs. If the purpose for now is jus to unblock internal use cases, private functions should be sufficient?

xush6528 replied:
We don't need these helpers, we only care about averages of all events.

The worker (usually a master) that enable this global profiler does not care about which worker put works on the server. It wants to get all profiling events on that server no matter which worker initiates the work.

mrshenli replied:
By "we" do you mean the current internal use case, or all future possible use cases including OSS?

xush6528 replied:
Current internal use case.

mrshenli reviewed (changes requested):
Thanks for adding this. I added some minor comments inline. 

My main concern is on the API usability. Looks like we would need more helpers + tests + examples + tutorials to make it easy to use and easy to understand. And after that we would need perf improve. If this is true, shall we make the two Python APIs private for now to unblock internal use cases and toggle them to public when all the missing gaps are filled?

xush6528 commented on torch/distributed/rpc/server_process_global_profiler.py:
```diff
@@ -0,0 +1,57 @@
+def disable_server_process_global_profiler():
+    """
+    Turn off the state that indicates the server-side process-global profiling is
+    on. Aggregrate all profiling events recorded by RPC threads.
```
Updating.

xush6528 commented on torch/distributed/rpc/server_process_global_profiler.py:
```diff
@@ -0,0 +1,57 @@
+    Returns:
+        event_list (torch.autograd.profiler.EventList). A list that have helper
+        methods like show record items in a pretty-print table,
+        do averaging by grouping on keys and more.
```
Yes, making them private API for now.

xush6528 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -578,6 +579,27 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processMessage(
+          // std::shared_lock. The cost is in magnitude
+          // of 10us. If server global profiler is enabled, we futher pay the
+          // cost of thread local profiler state initialization .
+          // int64_t startTime = ::torch::autograd::profiler::getTime();
```
Sorry. Removing.

xush6528 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -578,6 +579,27 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processMessage(
+          // int64_t startTime = ::torch::autograd::profiler::getTime();
+          bool isServerProcessGlobalProfilerEnabled =
+              profiler::processglobal::serverEnabled();
+          // int64_t endTime = ::torch::autograd::profiler::getTime();
+          // LOG(ERROR) << "Test profiler enabled took: "
+          //            << (endTime - startTime) / 1000 << " us";
```
Removing.

xush6528 commented on torch/distributed/rpc/server_process_global_profiler.py:
```diff
@@ -0,0 +1,57 @@
+
+    Returns:
+        None
+
```
Adding.

xush6528 committed:
Update on "[Profiler x RPC] Enable RPC Server Global Profiler"


See plan in https://github.com/pytorch/pytorch/issues/38845

Differential Revision: [D5665992](https://our.internmc.facebook.com/intern/diff/D5665992/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5665992/)!

[ghstack-poisoned]

mrshenli reviewed (approved):
Thanks for addressing comments. It LGTM to have those two APIs as private, and we can fill in gaps in followup PRs.

xush6528 committed:
Update on "[Profiler x RPC] Enable RPC Server Global Profiler"


See plan in https://github.com/pytorch/pytorch/issues/38845.

Close https://github.com/pytorch/pytorch/issues/38845.

Differential Revision: [D5665992](https://our.internmc.facebook.com/intern/diff/D5665992/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5665992/)!

[ghstack-poisoned]

xush6528 commented on torch/distributed/rpc/server_process_global_profiler.py:
```diff
@@ -0,0 +1,105 @@
+# ==================== Context API ====================
+
+
+class _server_process_global_profile(profile):
```
Update,

- Added a context-style python API. This will be used by fb internal.

rohan-varma commented on torch/distributed/rpc/server_process_global_profiler.py:
```diff
@@ -0,0 +1,105 @@
+
+class _server_process_global_profile(profile):
+    """
+    It has the same API as ``torch.autpgrad.profiler.profile`` class,
```
nit: spelling

xush6528 replied:
Thanks. Updating.

rohan-varma commented on torch/distributed/rpc/server_process_global_profiler.py:
```diff
@@ -0,0 +1,105 @@
+        methods like show record items in a pretty-print table,
+        do averaging by grouping on keys and more.
+    """
+    profile_ranges = __disable_server_process_global_profiler()
```
Is it possible to change the naming so that we don't have to prefix a function with two underscores?

xush6528 replied:
Updating the name to only have one underscore.

rohan-varma commented on torch/distributed/rpc/server_process_global_profiler.py:
```diff
@@ -0,0 +1,105 @@
+    """
+    It has the same API as ``torch.autpgrad.profiler.profile`` class,
+    except that it enables profiling on all threads running RPC server request callbacks.
+    """
```
Can you add an example of how to use this as added in the `rpc._enable_server_process_global_profiler` docstring? (As well as specify that

rohan-varma replied:
Also, if ` rpc._enable_server_process_global_profiler` isn't going to be meant to be used by users directly, adding a comment telling them to prefer using the context manager may be helpful.

xush6528 replied:
I am removing the functional API. As to be closer to `torch.autograd.profile`. I should only expose context style API.

rohan-varma commented on torch/distributed/rpc/server_process_global_profiler.py:
```diff
@@ -0,0 +1,105 @@
+        >>>     False,  # record_input_shapes
+        >>>     False,  # profile_memory
+        >>> )
+        >>> rpc.enable_server_process_global_profiler(profiler_config)
```
Is this the anticipated usage? If so, it looks different from the tests, where an RPC to enable it on the server is being sent.  Should the example be updated to send an RPC (or to the expected usage of the API)?

xush6528 replied:
Updating the example to be in a RPC setup.

rohan-varma commented on torch/distributed/rpc/server_process_global_profiler.py:
```diff
@@ -0,0 +1,105 @@
+        _enable_server_process_global_profiler(profiler_config)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
```
Is there a reason `__exit__`'s signature isn't the typical `__exit__(self, *args)`?

xush6528 replied:
I believe `def __exit__(self, exc_type, exc_val, exc_tb):` is more typical. 

`__exit__(self, *args)` is a lazy way of writing it, if the exec info is not used.

rohan-varma commented on torch/distributed/rpc/server_process_global_profiler.py:
```diff
@@ -0,0 +1,105 @@
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if not self.enabled:
+            return
```
Should we throw in this case?

xush6528 replied:
I am keeping the same logic as https://github.com/pytorch/pytorch/blob/235f62417d9d66b8f0bd62db385e8eba14a6c1ed/torch/autograd/profiler.py#L310-L311

rohan-varma reviewed (commented):
Left some comments/concerns that I have, but overall I like the idea of having a functional-oriented approach to enable the profiler across all server threads via a RPC call. Also, since this diff introduces some significant changes, it would be great to update the PR body with a description of the changes as well as how they're tested (the pointer to the RFC seems to be more of a high-level use case/plan)

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -976,6 +980,38 @@ def test_profiler_with_script_remote_rpc(self):
         )
 
+    @dist_init
+    def test_server_process_global_profiler(self):
```
Should we also test what happens in the case of double calls, calling exit without enter, calling exit twice, entering twice, etc?

xush6528 replied:
Adding.

xush6528 committed:
Update on "[Profiler x RPC] Enable RPC Server Global Profiler"


See plan in https://github.com/pytorch/pytorch/issues/38845.

Close https://github.com/pytorch/pytorch/issues/38845.

Differential Revision: [D5665992](https://our.internmc.facebook.com/intern/diff/D5665992/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5665992/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[Profiler x RPC] Enable RPC Server Global Profiler"


See plan in https://github.com/pytorch/pytorch/issues/38845.

Close https://github.com/pytorch/pytorch/issues/38845.

Differential Revision: [D5665992](https://our.internmc.facebook.com/intern/diff/D5665992/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5665992/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[Profiler x RPC] Enable RPC Server Global Profiler"


See plan in https://github.com/pytorch/pytorch/issues/38845.

Close https://github.com/pytorch/pytorch/issues/38845.

Differential Revision: [D5665992](https://our.internmc.facebook.com/intern/diff/D5665992/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D5665992/)!

[ghstack-poisoned]

xush6528 commented:
@rohan-varma Updated to resolve your comments.
